common/src/primitives/encoding/tests.rs

```diff
@@ -0,0 +1,264 @@
+            "split1checkupstagehandshakeupstreamerranterredcaperredlc445v",
+            "?1v759aa",
+        ).iter().for_each(|s| {
+            match super::bech32m::bech32m_to_base32(*s) {
+               Ok(decoded) => {
+                   match super::bech32m::base32_to_bech32m(decoded.get_hrp(), <&[u8]>::clone(&decoded.get_data())) {
+                       Ok(encoded) => { assert_eq!(s.to_lowercase(), encoded.to_lowercase()) }
+                       Err(e) => { panic!("Did not encode: {:?} Reason: {:?}",s,e) }
+                   }
+               }
+               Err(e) => {
+                   panic!("Did not decode: {:?} Reason: {:?}", s, e)
+               }
+           }
+        });
+}
+
+#[test]
+fn check_invalid_strings() {
+    vec!(
+            (" 1xj0phk", Bech32Error::InvalidChar(' ')),
+            ("\u{7F}1g6xzxy", Bech32Error::InvalidChar('\u{7f}')),
+            ("\u{80}1vctc34", Bech32Error::InvalidChar('Â')),
+            ("an84characterslonghumanreadablepartthatcontainsthetheexcludedcharactersbioandnumber11d6pts4", Bech32Error::InvalidLength),
+            ("qyrz8wqd2c9m", Bech32Error::NoSeparator),
+            ("1qyrz8wqd2c9m", Bech32Error::InvalidLength),
+            ("y1b0jsk6g", Bech32Error::InvalidChar('b')),
+            ("lt1igcx5c0", Bech32Error::InvalidChar('i')),
+            ("in1muywd", Bech32Error::InvalidLength),
+            ("mm1crxm3i", Bech32Error::InvalidChar('i')),
+            ("au1s5cgom", Bech32Error::InvalidChar('o')),
+            ("M1VUXWEZ", Bech32Error::FailedChecksum),
+            ("16plkw9", Bech32Error::InvalidLength),
+            ("1p2gdwpf", Bech32Error::InvalidLength),
+            ("bech321qqqsyrhqy2a", Bech32Error::UnsupportedVariant),
+            // invalid addresses
+            ("bcrt1q8p08mv8echkf3es027u4cdswxlylm3th76ls8v6y4zy4vwsavngpr4e4td", Bech32Error::UnsupportedVariant),
+            ("bc1q5cuatynjmk4szh40mmunszfzh7zrc5xm9w8ccy", Bech32Error::UnsupportedVariant),
+            ("bc1qkw7lz3ahms6e0ajv27mzh7g62tchjpmve4afc29u7w49tddydy2syv0087", Bech32Error::UnsupportedVariant),
+            ("tb1q74fxwnvhsue0l8wremgq66xzvn48jlc5zthsvz", Bech32Error::UnsupportedVariant),
+            ("tb1qpt7cqgq8ukv92dcraun9c3n0s3aswrt62vtv8nqmkfpa2tjfghesv9ln74", Bech32Error::UnsupportedVariant),
+            ("tb1q0sqzfp3zj42u0perxr6jahhu4y03uw4dypk6sc", Bech32Error::UnsupportedVariant),
+            ("tb1q9jv4qnawnuevqaeadn47gkq05ev78m4qg3zqejykdr9u0cm7yutq6gu5dj", Bech32Error::UnsupportedVariant),
+            ("bc1qz377zwe5awr68dnggengqx9vrjt05k98q3sw2n", Bech32Error::UnsupportedVariant),
+            ("tb1qgk665m2auw09rc7pqyf7aulcuhmatz9xqtr5mxew7zuysacaascqs9v0vn", Bech32Error::FailedChecksum)
+        ).iter().for_each(|(s,b_err)| {
+            match super::decode(*s) {
+                Ok(_) => { panic!("Should be invalid: {:?}", s) }
+                Err(e) => { assert_eq!(*b_err,e) }
+            }
+        });
```

Comment:
Please create a tests that:

1. Takes arbitrary u8 data, converts it to bech32, and then converts it back
2. Converts selected results from bech32 to expected values of arbitrary data, back and forth.